osu.Game/Screens/Play/Player.cs

```diff
@@ -632,6 +632,7 @@ public override void OnEntering(IScreen last)
 
             // bind component bindables.
             Background.IsBreakTime.BindTo(breakTracker.IsBreakTime);
+            gameplayBeatmap.IsBreakTime.BindTo(breakTracker.IsBreakTime);
```

Comment:
This is very questionable. `Player` should not be taking on logic to ensure that the beatmap is in the correct state.

I'd much rather that beatmap contain the tracker itself than do this. Is there a reason this can't be done?

Reply:
Haven't thought about that, makes sense by logic. will look at making `BreakTracker` part of the gameplay beatmap, but I'll open a separate PR as it may make things a bit complicated and add more irrelevant changes.

Reply:
Not sure about containing as the break tracker is placed in a frame stable context right now. I wouldn't mind making it DI-able, though.

Reply:
> I wouldn't mind making it DI-able, though.

@bdach Sounded good at first, but after attempting to do it, I realized it would become hard to test the combo counter's behaviour on break time, as I can't just set `IsBreakTime` to whatever value I want like I did preivously, I would be forced to hack through the `BreakTracker` with a manipulated clock or otherwise, which doesn't sound like a fine thing here.

---
As for merging `BreakTracker` to `GameplayBeatmap`, as mentioned above, the break tracker is placed in a frame stable context currently, moving the tracker to inside `GameplayBeatmap` would not be that easy.

At first, I thought about passing `DrawableRuleset` in and adding a private break tracker component inside their exposed frame stable components container, but that sounded shoddy mostly for the part that `GameplayBeatmap` would no longer own or be an ancestor to the private break tracker component, therefore could potentially result in unexpected issues.

I've then took a bit to think about how to do this properly, and I came up with the following solution: caching `GameplayBeatmap` right where it was previously, but without it being added to the hierarchy, and then load it after `DrawableRuleset` is loaded inside the frame stable components container where the score and health processors are added.

This will make `GameplayBeatmap` obtain the frame-stable clock (which is logical for a "gameplay beatmap"), allowing it to track breaks very easily, without having to change its dependency caching position.
 - Could cache `GameplayBeatmap` right where it is loaded, but it is possible not to change the cache position so I didn't do that, tell me if that's wrong to do.

@smoogipoo does that sound good?

Reply:
how would it be hard to test? can always expose a public `IsBreakTime` bindable on the combo counter (even if read-only), bind to it in the test and manipulate as required. there is precedent for this (see taiko mascot and scroller tests)

continuing that thread of thinking, you could also make the counter a `VisibilityContainer` or something.

Reply:
Because the test scene is dealing with `CatchComboDisplay`, not directly the legacy piece where the break logic resides. (`LegacyComboCounter`)

Not sure what you mean by making it a `VisibilityContainer`?

Reply:
Given the amount of complexity around this issue, I'm tempted to overlook it and fix it another day. Perhaps at some point we'll put a frame stable container inside the gameplay beatmap? Or inside player? Not sure...

@peppy what's your thoughts on leaving this as is?

Reply:
My view is that we should settle on anything because at this point it feels like overthinking such a simple component.

I'd apply my usual thinking: as long as it's isolated enough that it's not polluting other systems or components (more than a ctor and maybe a couple of property updates) code quality or structure is not important for a first pass.